packages/dd-trace/src/encode/agentless-ci-visibility.js

```diff
@@ -0,0 +1,245 @@
+'use strict'
+const tracerVersion = require('../../lib/version')
+const { truncateSpan, normalizeSpan } = require('./tags-processors')
+const Chunk = require('./chunk')
+const log = require('../log')
+const { AgentEncoder } = require('./0.4')
+
+const ENCODING_VERSION = 1
+
+function formatSpan (span) {
+  return {
+    type: span.type === 'test' ? 'test' : 'span',
+    version: ENCODING_VERSION,
+    content: normalizeSpan(truncateSpan(span))
+  }
+}
+
+class AgentlessCiVisibilityEncoder extends AgentEncoder {
+  constructor ({ runtimeId, service, env }) {
+    super(...arguments)
+    this._events = []
+    this.runtimeId = runtimeId
+    this.service = service
+    this.env = env
+    this._traceBytes = new Chunk()
+    this._stringBytes = new Chunk()
+    this._stringCount = 0
+    this._stringMap = {}
+
+    this.reset()
+  }
+
+  count () {
+    return this._events.length
+  }
+
+  append (trace) {
+    this._events = this._events.concat(trace)
+  }
+
+  _encodeValue (bytes, value) {
+    switch (typeof value) {
+      case 'string':
+        this._encodeString(bytes, value)
+        break
+      case 'number':
+        this._encodeNumber(bytes, value)
+        break
+      case 'object':
+        if (Array.isArray(value)) {
+          this._encodeArrayPrefix(bytes, value)
+          for (const event of value) {
+            this._encodeMap(bytes, event)
+          }
+          return
+        }
+        if (value.constructor && value.constructor.name === 'Identifier') {
+          this._encodeId(bytes, value)
+          return
+        }
+        this._encodeMap(bytes, value)
+        break
+      default:
+        // should not happen
+    }
+  }
+
+  _encodeNumber (bytes, value) {
+    if (Math.floor(value) !== value) { // float 64
+      return this._encodeFloat(bytes, value)
+    }
+    return this._encodeInteger(bytes, value)
+  }
+
+  _encodeInteger (bytes, value) {
+    if (value >= 0) {
+      this._encodePositiveLong(bytes, value)
+    } else {
+      this._encodeNegativeLong(bytes, value)
+    }
+  }
+
+  _encodeNegativeLong (bytes, value) {
```

Comment:
Why an explicit method? And why can't it be handled by the base encoder?

Reply:
I think I had trouble encoding negative integers (the intake was rejecting it). I might be wrong though. I'll give it another try and remove this function if it doesn't add anything

Reply:
so yeah it seems the distinction is needed. There might be a bug in 0.4 for negative integers

Reply:
yeah the distinction seems to be needed, otherwise integers are wrong. I think it works for 0.4 because every number is encoded as a float, but that doesn't work for agentless

Reply:
I can't seem to find this method anymore, did you end up doing this another way?

Reply:
I did it in this commit: https://github.com/DataDog/dd-trace-js/pull/1880/commits/3020ffda5b738d3b78170b19745f4621c3039b51